### PR TITLE
sl-10: Sign up

### PR DESCRIPTION
**Pull request by dimapopovych**
image

### CODE REVIEW

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts [hunk 1]
```diff
@@ -3,21 +3,39 @@ import joi from 'joi';
...
+  /^(\w)([\w+.-]{0,34}\w)?@[\dA-Za-z][\dA-Za-z-]{0,61}[\dA-Za-z]\.[A-Za-z]{2,}$/;
+const PASSWORD_REGEX = /^[\d!"#$%&()-_`a-z{|}~]{8,30}$/;
+const FIRST_NAME_REGEX = /^['A-Za-z-]+$/;
+const LAST_NAME_REGEX = /^['A-Za-z-]+$/;
+
 const userSignUp = joi.object<UserSignUpRequestDto, true>({
   email: joi
     .string()
-    .trim()
+    .regex(EMAIL_REGEX)
```
Comment (what1s1ove) [inline shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]:
```suggestion
    .regex(UserValidationRule.EMAIL_REGEX)
```

File: frontend/src/assets/css/scaffolding.scss [hunk 2]
```diff
@@ -9,7 +9,7 @@ body {
   min-height: 100vh;
   margin: 0;
   font-size: 14px;
-  font-family: Arial, sans-serif;
+  font-family: Roboto, sans-serif;
```
Comment (what1s1ove) [inline frontend/src/assets/css/scaffolding.scss:12]: it should be added to the index.html

File: frontend/src/assets/css/styles.scss [hunk 3]
```diff
@@ -1 +1,3 @@
 @import url("./scaffolding.scss");
+@import url("./variables.scss");
+@import url("./styles.text.scss");
```
Comment (what1s1ove) [inline frontend/src/assets/css/styles.scss]:
```suggestion
```
only variables should be global

File: frontend/src/libs/components/button/button.tsx [hunk 4]
```diff
@@ -1,11 +1,35 @@
+import './styles.scss';
```
Comment (what1s1ove) [inline frontend/src/libs/components/button/button.tsx]: use style modules

File: frontend/src/libs/components/button/button.tsx [hunk 5]
```diff
@@ -1,11 +1,35 @@
...
+
 type Properties = {
   label: string;
   type?: 'button' | 'submit';
+  isFull?: boolean;
+  isSmall?: boolean;
```
Comment (what1s1ove) [inline frontend/src/libs/components/button/button.tsx]:
```suggestion
```
common component should have 100% of the width

File: frontend/src/libs/components/input/input.tsx [hunk 6]
```diff
@@ -14,6 +16,7 @@ type Properties<T extends FieldValues> = {
   name: FieldPath<T>;
   placeholder?: string;
   type?: 'text' | 'email';
+  disabled?: boolean;
```
Comment (what1s1ove) [inline frontend/src/libs/components/input/input.tsx]: check *B7*

File: frontend/src/libs/components/input/input.tsx [hunk 7]
```diff
@@ -1,3 +1,5 @@
+import './styles.scss';
```
Comment (what1s1ove) [inline frontend/src/libs/components/input/input.tsx]: use style modules

**Review by what1s1ove — CHANGES_REQUESTED** [review]
frontend/src/assets/img/background-pattern.png - should be svg i guess

File: frontend/src/assets/css/scaffolding.scss (hunk 2 above)
Comment (dimapopovych) [reply frontend/src/assets/css/scaffolding.scss:12]: I don't quite get it. Am I supposed to add style tag to intex.html and here specify font-family?

**Comment by dimapopovych** [comment]
> frontend/src/assets/img/background-pattern.png - should be svg i guess

Simply importing this image as svg from Figma doesn't work. File size of svg 4x bigger then png.

**Comment by dimapopovych** [comment]
Screenshot 2023-03-14 191946
After changes

File: frontend/index.html [hunk 8]
```diff
@@ -3,6 +3,9 @@
   <head>
     <meta charset="UTF-8" />
     <link rel="icon" type="image/svg+xml" href="/vite.svg" />
+    <link rel="preconnect" href="https://fonts.googleapis.com" />
+    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
```
Comment (what1s1ove) [inline frontend/index.html]:
```suggestion
```

File: frontend/index.html [hunk 9]
```diff
@@ -3,6 +3,9 @@
...
+    <link href="https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap" rel="stylesheet" />
```
Comment (what1s1ove) [inline frontend/index.html]: not sure that you need all of these fonts. select only we needed

File: frontend/src/libs/components/input/input.tsx [hunk 10]
```diff
@@ -7,22 +7,21 @@ import {
...
+
 type Properties<T extends FieldValues> = {
   control: Control<T, null>;
   errors: FieldErrors<T>;
   label: string;
   name: FieldPath<T>;
-  placeholder?: string;
-  type?: 'text' | 'email';
-};
+} & React.ComponentPropsWithoutRef<'input'>;
```
Comment (what1s1ove) [inline frontend/src/libs/components/input/input.tsx]:
```suggestion
};
```
provide props that you need

File: frontend/src/libs/components/button/button.tsx [hunk 11]
```diff
@@ -1,11 +1,25 @@
+import styles from './styles.module.scss';
+
 type Properties = {
   label: string;
-  type?: 'button' | 'submit';
-};
+  size?: 'big' | 'small';
+  style?: 'primary' | 'secondary' | 'plain';
+  className?: string;
+} & React.ComponentPropsWithoutRef<'button'>;
```
Comment (what1s1ove) [inline frontend/src/libs/components/button/button.tsx]:
```suggestion
};
```

File: frontend/src/libs/components/button/button.tsx [hunk 12]
```diff
@@ -1,11 +1,25 @@
...
   label,
-}: Properties) => <button type={type}>{label}</button>;
+  size = 'big',
+  style = 'primary',
+  className = '',
+  ...mirroredProperties
+}: Properties) => (
+  <button
+    className={`${styles.button} ${styles[style]} ${styles[size]} ${className}`}
+    {...mirroredProperties}
```
Comment (what1s1ove) [inline frontend/src/libs/components/button/button.tsx]:
```suggestion
```

File: shared/src/packages/users/libs/enums/user-validation-rules.enum.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+const UserValidationRules = {
```
Comment (what1s1ove) [inline shared/src/packages/users/libs/enums/user-validation-rules.enum.ts]: check **B12**

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

File: frontend/src/libs/components/input/input.tsx [hunk 14]
```diff
@@ -31,9 +35,15 @@ const Input = <T extends FieldValues>({
 
   return (
     <label>
-      <span>{label}</span>
-      <input {...field} type={type} placeholder={placeholder} />
-      {hasError && <span>{error as string}</span>}
+      <span className={styles.inputLabel}>{label}</span>
+      <input
+        {...field}
+        className={`${styles.input} ${hasError ? styles.hasError : ''}`}
```
Comment (andanf-e) [inline frontend/src/libs/components/input/input.tsx]: Please check [this comment](https://github.com/BinaryStudioAcademy/bsa-winter-2022-2023-simpleland/pull/27#discussion_r1135892439) (about multiple/conditional classnames)

File: package.json [hunk 15]
```diff
@@ -48,5 +48,8 @@
   "simple-git-hooks": {
     "pre-commit": "npx lint-staged",
     "commit-msg": "npx commitlint --edit $1"
+  },
+  "dependencies": {
+    "clsx": "1.2.1"
```
Comment (andanf-e) [inline package.json]: Guess it should be frontend dependency
```suggestion

```

File: frontend/src/libs/components/input/input.tsx [hunk 16]
```diff
@@ -1,3 +1,4 @@
+import { clsx } from 'clsx';
```
Comment (andanf-e) [inline frontend/src/libs/components/input/input.tsx]: Let's import clsx only in one place (create smth like `getValidClassnames` helper and use it)

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/libs/components/input/input.tsx [hunk 17]
```diff
@@ -31,9 +36,15 @@ const Input = <T extends FieldValues>({
...
-      {hasError && <span>{error as string}</span>}
+      <span className={styles.inputLabel}>{label}</span>
+      <input
+        {...field}
+        className={clsx(styles.input, hasError && styles.hasError)}
+        type={type}
+        placeholder={placeholder}
+        disabled={isDisabled}
+      />
+      <div className={styles.errorMessage}>{hasError && (error as string)}</div>
```
Comment (what1s1ove) [inline frontend/src/libs/components/input/input.tsx]:
```suggestion
      <span className={styles.errorMessage}>{hasError && (error as string)}</span>
```
`label` cannot contain `div`. it should be a span

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: package.json (hunk 15 above)
Comment (dimapopovych) [reply package.json]: fixed

File: frontend/src/libs/components/input/input.tsx (hunk 16 above)
Comment (dimapopovych) [reply frontend/src/libs/components/input/input.tsx]: fixed

Comment (dimapopovych) [reply frontend/src/libs/components/input/input.tsx]: fixed

File: frontend/src/libs/components/input/input.tsx (hunk 17 above)
Comment (dimapopovych) [reply frontend/src/libs/components/input/input.tsx]: But I need to specify height for errorMessage. Can I use span with display: block property?

Comment (what1s1ove) [reply frontend/src/libs/components/input/input.tsx]: yes, you can

Comment (dimapopovych) [reply frontend/src/libs/components/input/input.tsx]: fixed

**Review by what1s1ove — APPROVED** [review]

File: frontend/src/libs/components/page-layout/page-layout.tsx [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+type Properties = {
+  style?: 'yellow' | 'black';
+  children: React.ReactNode;
+};
+
+const PageLayout: React.FC<Properties> = ({
+  children,
+  style = 'yellow',
+}: Properties) => (
+  <div className={`${styles.pageLayout} ${styles[style]}`}>{children}</div>
```
Comment (andanf-e) [inline frontend/src/libs/components/page-layout/page-layout.tsx]: FYI about `getValidClassnames` usage

File: frontend/src/assets/css/styles.scss (hunk 3 above)
Comment (Alyona-Pisotska) [reply frontend/src/assets/css/styles.scss]: @what1s1ove можна уточнення щодо цього глобального файлу. я його зробила - там всі стилі для кожного виду тексту, щоб можна було глобально використати клас, а не дублювати для кожного блоку різного тексту свої властивості. це помилково?

File: frontend/src/libs/components/page-layout/page-layout.tsx (hunk 18 above)
Comment (dimapopovych) [reply frontend/src/libs/components/page-layout/page-layout.tsx]: forgot that one, fixed

**Review by andanf-e — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**